### DISCUSSION TITLE
SAX issues / improvement

### DISCUSSION THREAD

**Discussion by flyx** [nlohmann/json#2275] [Ideas]
Seeing that there is a SAX API for loading JSON, I wonder whether a similar API for generating JSON would make sense. If I want to serialize some struct to JSON, usually I would not care about an intermediate in-memory JSON representation and directly generating SAX events to build the JSON from seems to be more efficient.

One problem with the current SAX API is that unlike `from_json` / `to_json` it cannot easily be used with composite types. This would be different if the SAX API provided an object from which the next event could be repeatedly queried. You could then have functions `from_json_stream`, and for a composite type they delegate the parsing of child objects to the `from_json_stream` of the child's type. Those functions could be generated via macros similar to `from_json`, and a reverse function `to_json_stream` could be generated if there was a SAX generator API.

Thoughts?

**Comment by nlohmann** [comment]
Let me see if I understood correctly. So given a class `widget`, you would propose having a function `void to_json_stream(const widget& w, sax_writer& sax)` in which you can define which SAX events serializing a `widget` would trigger. For instance

```cpp
class widget {
  public:
    int i;
    std::string s;
};

void to_json_stream(const widget& w, sax_writer& sax) {
    sax.start_object(2);
    sax.key("i");
    sax.number_integer(w.i);
    sax.key("s");
    sax.string(w.s);
    sax.end_object();
}
```

And then an implementation of `sax_writer` can decide what to do about this, for instance writing `{"i":1,"s":"Hello"}` to some stream without ever holding the whole JSON value in memory?

**Reply by flyx** [reply]
Exactly. To highlight how this can be autogenerated with a macro, the values would generally use `to_json_stream`:

```c++
void to_json_stream(const widget& w, sax_writer& sax) {
    sax.start_object(2);
    sax.key("i");
    to_json_stream(w.i, sax);
    sax.key("s");
    to_json_stream(w.s, sax);
    sax.end_object();
}
```

**Reply by nlohmann** [reply]
I like the idea, because it would also allow to create other formats like CBOR on the fly.

There is a lot to do, though. And one issue I struggled when using the SAX interface for internal parsing to a DOM was the fact that you need to hold some state. But I think it's worth investigating whether using the SAX interface is already worthwhile for the current serializers - regardless of user-defined types.

**Reply by flyx** [reply]
I basically implemented this for [NimYAML](https://nimyaml.org/) and it works well there. What I did was to make an event stream interface (so, basically SAX) the central processing component. Both parser and serializer implemented the event stream interface (transforming character input / data types into an event stream respectively) and deserializer / presenter consumed it creating data types / character output.

Interestingly, this library's API maps quite well to the loading process described by YAML, which was the base for my architecture:

![](https://yaml.org/spec/1.2/overview2.png)

A `json` object would be a **Representation** and `to_json` / `from_json` would be *Represent* / *Construct*. The SAX API would be the **Serialization**.

Not sure if this helps, your architecture might be quite different.

**Reply by nlohmann** [reply]
Thanks for the image - it is really helpful!

**Comment by flyx** [comment]
Another thought: Having a SAX API that allowed `from_json_stream` would, unlike current macro-generated `from_json` implementations, be able to detect unknown fields in the JSON, which are silently ignored in autogenerated `from_json`. Like this:

```c++
class widget {
  public:
    int i;
    std::string s;
};

void from_json_stream(widget &w, sax_reader &r) {
  static std::string i("i");
  static std::string s("s");
  if (r.next() != sax_event::object_start) throw exception(42, "expected object start, got …");
  for (sax_event cur = r.next(); cur != sax_event::object_end; cur = r.next()) {
    assert(cur == sax_event::key);
    if (r.str_content() == i) {
      from_json_stream(w.i, r);
      continue;
    }
    if (r.str_content() == s) {
      from_json_stream(w.s, r);
      continue;
    }
    throw exception(47, "unknown field: …");
  }
}
```

This `from_json_stream` could be created with a macro just like `from_string`.

**Comment by garethsb** [comment]
I implemented something similar around another JSON library, including plain and pretty-printed serialisation and DOM construction. It's been very useful in avoiding constructing large in-memory objects, in filtering out properties during serialisation, doing on-the-fly HTML rendering and the like.

Interface (template based visitor pattern) and the basic implementations (which demonstrate how much state is required) are here - https://github.com/sony/nmos-cpp/blob/master/Development/cpprest/json_visit.h

**Reply by nlohmann** [reply]
Thanks for sharing! I am currently struggling with building something similar, see #2284.

**Comment by Ericson2314** [comment]
I had commented 2 years ago on https://github.com/nlohmann/json/discussions/2284, but I think this discussion is actually more appropriate, because what we needed for https://github.com/NixOS/nix/pull/3893 is SAX-like printing, and doing that while dogfooding it in the regular printing code is much easier.

(two different ways to print things does not induce the problems that two different ways to parse things does)
